### PR TITLE
[master < T0711-MG] Add support for temporal types in pymgclient

### PR DESCRIPTION
**Pull request by kostasrim**
CI fails because `T0712-MG` needs to get merged first.

### CODE REVIEW

**Comment by kostasrim** [comment]
We should drop the updated ref of mgclient before we merge.

File: src/glue.c [hunk 1]
```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
...
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, div_result, NULL);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(result, NULL);
+  SCOPED_CLEANUP PyObject *h = PyObject_GetAttrString(result, "hour");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(h, NULL)
+  SCOPED_CLEANUP PyObject *m = PyObject_GetAttrString(result, "minute");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(m, NULL)
+  SCOPED_CLEANUP PyObject *s = PyObject_GetAttrString(result, "second");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(s, NULL)
+  SCOPED_CLEANUP PyObject *mi = PyObject_GetAttrString(result, "microsecond");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(mi, NULL)
```
Comment (antonio2368) [inline src/glue.c:311]: Why not just divide and mod for all the values? Not sure why use unix_epoch for LocalTime where the epoch is midnight. Also, don't we lose on precision with the floating point numbers?

File: src/glue.c [hunk 2]
```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
...
+      PyDateTime_FromDateAndTime(1970, 1, 1, 0, 0, 0, 0);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(unix_epoch, NULL)
+  SCOPED_CLEANUP PyObject *seconds =
+      PyLong_FromLong(mg_local_date_time_seconds(ldt));
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(seconds, NULL)
+  SCOPED_CLEANUP PyObject *method_name =
+      PyUnicode_FromString("utcfromtimestamp");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(method_name, NULL)
+  SCOPED_CLEANUP PyObject *result =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, seconds, NULL);
```
Comment (antaljanosbenjamin) [inline src/glue.c:328]: If you use [`PyObject_CallMethod`](https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethod) then you don't need to create a `PyUnicode`.

The same true for all other occurrence of `PyObject_CallMethodObjArgs`.

File: src/glue.c [hunk 3]
```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
...
+  SCOPED_CLEANUP PyObject *denominator = PyLong_FromLong(1000000000);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(denominator, NULL)
+  SCOPED_CLEANUP PyObject *div_result =
+      PyNumber_TrueDivide(numerator, denominator);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(div_result, NULL)
+  SCOPED_CLEANUP PyObject *method_name =
+      PyUnicode_FromString("utcfromtimestamp");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(method_name, NULL)
+  SCOPED_CLEANUP PyObject *result =
+      PyObject_CallMethodObjArgs(unix_epoch, method_name, div_result, NULL);
```
Comment (antaljanosbenjamin) [inline src/glue.c:302]: I think you are calling a [class method](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp) on a concrete instance of a `datetime` object. I think you should get `datetime.datetime` type as a `PyObject` and call the `utcfromtimestamp` on that PyObject and not on a concrete instance of `datetime.datetime`.

I saw this multiple time, but I won't comment on all of them.

File: src/glue.c [hunk 4]
```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
...
+  SCOPED_CLEANUP PyObject *y = PyObject_GetAttrString(result, "year");
+  SCOPED_CLEANUP PyObject *mo = PyObject_GetAttrString(result, "month");
+  SCOPED_CLEANUP PyObject *d = PyObject_GetAttrString(result, "day");
+  SCOPED_CLEANUP PyObject *h = PyObject_GetAttrString(result, "hour");
+  SCOPED_CLEANUP PyObject *m = PyObject_GetAttrString(result, "minute");
+  SCOPED_CLEANUP PyObject *s = PyObject_GetAttrString(result, "second");
+  int64_t nanos = mg_local_date_time_nanoseconds(ldt);
+  return make_py_datetime(PyLong_AsLong(y), PyLong_AsLong(mo), PyLong_AsLong(d),
+                          PyLong_AsLong(h), PyLong_AsLong(m), PyLong_AsLong(s),
+                          (nanos / 1000));
```
Comment (antaljanosbenjamin) [inline src/glue.c:339]: Isn't `result` already a `datetime` object? I think there is no need to get the member of it and then reassemble it in `make_py_datetime`.

File: src/glue.c [hunk 5]
```diff
@@ -349,7 +489,113 @@ mg_map *py_dict_to_mg_map(PyObject *dict) {
...
+  SCOPED_CLEANUP PyObject *method_name = PyUnicode_FromString("__sub__");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(method_name, 0)
+  SCOPED_CLEANUP PyObject *delta =
+      PyObject_CallMethodObjArgs(date, method_name, unix_epoch, NULL);
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(delta, 0)
+  SCOPED_CLEANUP PyObject *days = PyObject_GetAttrString(delta, "days");
+  IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(days, 0)
+  *result = PyLong_AsLong(days);
+  return 1;
+}
```
Comment (antaljanosbenjamin) [inline src/glue.c:509]: It feels strange to call python for this. But I couldn't figure out how it can be done in C.

File: src/glue.c [hunk 6]
```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
...
+void maybe_decrement_ref(PyObject **obj) {
+  if (*obj) {
+    Py_DECREF(obj);
+  }
+}
+
+#define SCOPED_CLEANUP __attribute__((cleanup(maybe_decrement_ref)))
+#define IF_PTR_IS_NULL_PRINT_ERROR_AND_RETURN(ptr, value) \
+  if (!(ptr)) {                                           \
+    PyErr_Print();                                        \
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/glue.c:229]: Yes, upon discussing this offline, if we print the error the user will never get an exception. Fixed.

File: src/glue.c (hunk 2 above)
Comment (kostasrim) [reply src/glue.c:328]: Yes but it has other downsides that are describe here: `https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethod` 

1. For the C arguments I have to use `The C arguments are described by a Py_BuildValue()`
2. `Note that if you only pass PyObject * args, PyObject_CallMethodObjArgs() is a faster alternative.`

File: src/glue.c (hunk 5 above)
Comment (kostasrim) [reply src/glue.c:509]: Yes because it's so much easier to do this in Python even if you have to go through the C wrappers. One of the reasons I was talking about WASM offline ;)

File: src/glue.c (hunk 4 above)
Comment (kostasrim) [reply src/glue.c:339]: Good point but there is a catch. `datetime` objects are immutable so you can't call `SetAttrString()` to set the `microseconds` field. Hence you either need to call `[replace](https://docs.python.org/3/library/datetime.html#datetime.datetime.replace)` or `make_py_datetime`. I prefer the second because I already have this utility function and I don't really like calling python methods from C.

File: src/glue.c [hunk 7]
```diff
@@ -214,6 +217,135 @@ PyObject *mg_path_to_py_path(const mg_path *path) {
...
+  PyObject *time = PyTime_FromTime(h, min, sec, mi);
+  if (!time) {
+    PyErr_Print();
+  }
+  return time;
+}
+
+PyObject *make_py_datetime(int y, int m, int d, int h, int min, int sec,
+                           int mi) {
+  PyDateTime_IMPORT;
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/glue.c:253]: This defines a static variable which is then used by each of datetime objects. Therefore, it either must be put in a global scope such that all instances have access to this function, or the way it is done now. I am unsure which one is the best.

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/glue.c:253]: https://github.com/python/cpython/blob/main/Include/datetime.h#L200

[the remark this answers is not in the document]
Comment (antonio2368) [reply src/glue.c:253]: yes, the initialized variable is static so I presume that calling this function in `mgclientmodule.c` that is a different TU won't initialize the correct object for this TU. I would also like to avoid all the calls to import but I'm not sure how to achieve this.

[the remark this answers is not in the document]
Comment (kostasrim) [reply src/glue.c:253]: I don't know either, if we move it to `mgclientmodule.c` is segfaults (because of the way it's defined look at the above link)

File: src/glue.c (hunk 3 above)
Comment (kostasrim) [reply src/glue.c:302]: I have been trying to figure this out but I don't know. It seems like, that if you convert `PyObjectType` to a `PyObject` you are actually creating an instance of its class. I could (potentially) reach the static method somehow through its [structure](https://github.com/python/cpython/blob/main/Include/datetime.h#L200) but this seems extremely convoluted. On the other hand, you can always call a noninstance method from an object of this class without an issue which is basically what is being done in the code above with the `unix_epoch`. Moreover, `CallMethod*` requires `PyObject` which implies an instance of an object (because a PyObject is just a wrapper for a PyObjectType + a reference count). I will investigate more and try to figure a way to do this. Maybe you have an idea?

Comment (antonio2368) [reply src/glue.c:302]: Could you use `PyDateTime_FromTimestamp` and `PyDateTime_TimeZone_UTC`?

File: src/glue.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (antaljanosbenjamin) [reply src/glue.c:252]: Ahh, I didn't know it is static. Then the current status is okay for me.

File: src/glue.c (hunk 2 above)
Comment (antaljanosbenjamin) [reply src/glue.c]: 1. That's true, but I think that's fine as this is how Python API works.
2. That might be true, but when you create the `PyUnicode` it will copy all the bytes, allocate etc.

What if we define these values for global object for this file (or globally in the whole module)? Then we can use `PyObject_CallMethodObjArgs` without copying the same bytes over and over again. Or maybe create a static variable inside the function, though I am not sure if that is possible in C or not.

File: src/glue.c (hunk 4 above)
Comment (antaljanosbenjamin) [reply src/glue.c]: So here we only use python to convert the seconds to the appropriate values. Then I think this is fine.

File: src/glue.c (hunk 2 above)
Comment (kostasrim) [reply src/glue.c]: 1. Yes but if I have an alternative why would I not choose that? They are both API functions.
2. Wouldn't it be the same case with args? You, still need to allocate on these cases as well ?
I think you can use static variables in a function but this creates a new problem. Who owns the lifetime of the string? Unfortunately, `SCOPED_CLEANUP` works only for  automatic storage duration.

We decided offline to check alternatives and we will update this post :)

File: src/glue.c (hunk 3 above)
Comment (kostasrim) [reply src/glue.c]: @antonio2368 I could use  `PyDateTime_FromTimestamp`. However, I have to convert  the units to microseconds +  allocate the `PyObject` which is a little bit cumbersome.

**Review by jbajic — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -8,9 +8,9 @@ jobs:
      matrix:
         platform: [ubuntu-18.04]
         mgversion:
-          - 1.2.0
           - 1.3.0
           - 1.4.0
+          - 2.0.0
```
Comment (antaljanosbenjamin) [inline .github/workflows/ci.yml:13]: Maybe add `1.6.1` also.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -23,7 +23,11 @@ jobs:
       if: steps.cache-memgraph-community.outputs.cache-hit != 'true'
       run: |
         mkdir ~/memgraph
-        curl -L https://download.memgraph.com/memgraph/v${{ matrix.mgversion }}/ubuntu-18.04/memgraph_${{ matrix.mgversion }}-community-1_amd64.deb > ~/memgraph/memgraph-community.deb
+        memgraph_download_url=https://download.memgraph.com/memgraph/v${{ matrix.mgversion }}/ubuntu-18.04/memgraph_${{ matrix.mgversion }}-community-1_amd64.deb
+        if [ ${{ matrix.mgversion }} == "2.0.0" ]; then
+          memgraph_download_url=https://download.memgraph.com/memgraph/v${{ matrix.mgversion }}/ubuntu-18.04/memgraph_${{ matrix.mgversion }}-1_amd64.deb
+        fi
```
Comment (antaljanosbenjamin) [inline .github/workflows/ci.yml:29]: What if only the `-community` part would depend on the version?

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -39,7 +43,13 @@ jobs:
     - name: Install pymgclient
       run: python3 -m pip install ./dist/pymgclient-*
     - name: Run tests
-      run: MEMGRAPH_PORT=10000 python3 -m pytest
+      run: |
+        MEMGRAPH_PORT=10000
+        if [ ${{ matrix.mgversion }} == "2.0.0" ]; then
```
Comment (antaljanosbenjamin) [inline .github/workflows/ci.yml:48]: Maybe use `>=` to make sure temporal type tests will run also in the future versions? Also below in the windows run.

File: test/test_glue.py [hunk 11]
```diff
@@ -228,32 +228,38 @@ def test_tuple(memgraph_connection):
...
-
-# def test_duration(memgraph_connection):
-#     conn = memgraph_connection
-#     cursor = conn.cursor()
-#     cursor.execute("RETURN $value", {'value': datetime.timedelta(64, 7, 11, 1)})
-#     result = cursor.fetchall()
-#     assert result == [(datetime.timedelta(64, 7, 1011),)]
+
+@pytest.mark.temporal
+def test_time(memgraph_connection):
```
Comment (antaljanosbenjamin) [inline test/test_glue.py:233]: From our experiences, it might be good to add some tests to negative values.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -5,57 +5,68 @@ on: [push]
...
+        with:
+          submodules: true
+      - name: Build source distribution
+        run: python3 setup.py sdist
+      - name: Install pymgclient
+        run: python3 -m pip install ./dist/pymgclient-*
+      - name: Run tests
+        run: |
+          MEMGRAPH_PORT=10000
+          if [ ${{ matrix.mgversion }} == "2.0.0" ]; then
```
Comment (antaljanosbenjamin) [inline .github/workflows/ci.yml:49]: This still needs to be improved.

**Review by antonio2368 — APPROVED** [review]
